Fixed #20147 -- Added request.headers.

santiagobasulto opened the pull request:
Hello 👋! My first PR, so please double check my code 😁.

#### Background

Django’s primary deployment platform is [WSGI](https://www.python.org/dev/peps/pep-3333/). WSGI uses an [`environ`](https://www.python.org/dev/peps/pep-3333/#environ-variables) dictionary similarly to CGI, which converts all headers to uppercase, replaces dashes with underscores and prepends `HTTP_`. For example, if the client sends the header `If-None-Match`, WSGI would store it in `environ` as `HTTP_IF_NONE_MATCH`.

This makes the process of accessing familiar headers a little bit more cumbersome. The ticket #20147 proposes making original headers more easily accesible.

#### Proposed Solution

My solution is simple. Now `requests` has a `headers` attribute which is an immutable, case-insensitive dictionary with the original names of the headers. I decided `request.headers` because is the most intuitive way of thinking about it, and case insensitive because HTTP specifies so: _"Field names are case-insensitive"_ (see [RFC](https://tools.ietf.org/html/rfc2616#page-31)). Example of proposed solution:

```python
def my_view(request):
    if 'Accept' in request.headers or 'referer' in request.headers:
        pass
    cache_control = request.get('Cache-Control', None)
    accept_encoding = request.headers['Accept-Encoding']
    connection = request.headers.get('connection')  # case insensitive
```

#### Implementation details

I looked for "inspiration" (😬) in [Werkzeug](https://github.com/pallets/werkzeug) (the wsgi framework behind Flask). They implement an [`EnvironHeaders`](https://github.com/pallets/werkzeug/blob/43fe889fe5a8c495c41ec28658f67c1f7b23e2cd/werkzeug/datastructures.py#L1326) which receives the `environ` dict and transforms headers to the original format (assuming title case). But, I didn't implement the dictionary features in my `EnvironHeaders`. Instead I separated it into another data structure `ImmutableCaseInsensitiveDict` inspired by requests'  [CaseInsensitiveDict](https://github.com/requests/requests/blob/239ee194318e0b88d6a7b292ec4b68451ceefe68/requests/structures.py#L15) (more "inspiration" 😅).

#### Design Decisions

A few design decisions that I arbitrarily made and are obviously subject to discussion:

* As mentioned above, separated `EnvironHeaders` from the underlying data structure because we might be able to reuse it later.
* I didn't modify the current way the framework uses headers. For example, `GZipMiddleware` is still accessing `request.META`. It wouldn't be hard to trace down all the headers being accessed through `request.META` and change them to `request.headers`, but I figured I could create **many** merge conflicts by doing so. So please advice.
* I updated the docs accordingly, but didn't indicate any preference or recommendation of one over the other. 

I repeat, this is my first PR, so comments, suggestion, criticism are welcome.

carltongibson commented on docs/ref/request-response.txt:
```diff
@@ -159,14 +162,35 @@ All attributes should be considered read-only, unless stated otherwise.
+    Header case is preserved when turning it into a dictionary,
+    or iterating over it (you'll see the original header sent).
+    But when accessing elements with ``headers.get``, ``headers[]`` or
+    checking membership (``in``), case is ignored. Examples::
```
Can you explain this? (I'm not sure I quite follow it.) 

* Not matter what the client sends, WSGI gives us headers of the form `HTTP_USER_AGENT` right? 
* These are then transformed in `parse_cgi_header` to hyphened-title-case (is there a name for that?) such as, `User-Agent` (`return cgi_header.replace('_', '-').title()`)
* So in what sense is header case preserved or we "see the original header sent"?

santiagobasulto replied:
You're totally right. I didn't choose the correct words.

carltongibson commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -155,3 +155,230 @@ def f(x):
+        self.assertTrue('Content-Type' in self.dict_1)
+
+    def test_delitem(self):
+        "del should raise a TypeError because this is immutable"
```
Please use `"""` for docstrings

carltongibson commented on docs/ref/request-response.txt:
```diff
@@ -159,14 +162,35 @@ All attributes should be considered read-only, unless stated otherwise.
+
+            # .get():
+            request.headers.get('Accept')  # Returns the value of Accept
+            request.headers.get('accept')  # Returns the value of Accept
```
There should probably be an example of a multi-word header, such as `User-Agent`, here.

carltongibson commented on docs/ref/request-response.txt:
```diff
@@ -135,8 +135,11 @@ All attributes should be considered read-only, unless stated otherwise.
+    A dictionary containing the `WSGI environ \
+    <https://www.python.org/dev/peps/pep-0333/#environ-variables>`_.
+    Available request headers are prefixed with the string ``HTTP_``.
+    If you're looking for the original header names sent by the client,
```
Not sure about "original header names sent by the client"

spookylukey commented:
This looks like a good start, thanks so much.

A key issue with this feature is unicode and what happens if header values contain non-ascii chars. For that we need to think:
1. Are there any relevant RFCs for this? Based on a comment [here](http://lucumr.pocoo.org/2010/5/25/wsgi-on-python-3/) perhaps headers need to assume latin1 encoding always.
2. What exactly does WSGI do to the incoming bytes? https://www.python.org/dev/peps/pep-3333/#unicode-issues
3. Do we need to apply DEFAULT_CHARSET at any point?
4. Can we get close to an 'end to end' test for non-ascii characters to ensure the right thing happens, i.e. start with a set of bytes that would arrive, and check what happens to them?

Also be aware that in various docs 'headers' might refer to response OR request headers, here we are obviously only talking about the latter.

The good news is that we are Python 3 only for this feature, which should make things simpler.

Sorry to just throw in a bunch of questions, I won't be around to give further feedback.

collinanderson commented:
- Re lower-case vs Titled-Hyphens, I don't have a strong opinion at this point. Either's fine with me. Titled-Hyphens has seemed to stay pretty standard, despite HTTP2 taking off.

- Re unicode, I believe wsgi is already handling this (because we're on python 3). I think we're already getting unicode strings.

santiagobasulto commented:
Thanks very much folks @carltongibson, @spookylukey, @collinanderson. I can make all the requested changes. I'd like to summarize them first to see if everybody else agree on them:

#### Lowercase Headers
As @carltongibson mentioned, HTTP/2 [specifies that they must be converted to lowercase](https://tools.ietf.org/html/rfc7540#section-8.1.2). I can do that. After all, it's a case-insensitive dict, the `title` vs `lower` difference is just shown in `repr`, `str` or `dict` (I guess when you're debugging the app). The user can still use `headers.get('User-Agent')` if they prefer so.

**Conclusion:** I can quickly change it. Should I? 👍/👎

#### Headers Encoding

This is a tough one. According to the RFC, headers can't carry characters outside latin1. I did some research and found [an interesting comment by @Lukasa with more details](https://github.com/requests/requests/issues/1926#issuecomment-35524028) in the requests project. Also some [discussion (more interesting comments)](https://github.com/jakubroztocil/httpie/issues/212) in the httpie project. Python 3 doesn't even support **sending** non-latin1 headers. They [mention RFC2616](https://github.com/python/cpython/blob/b00854caa080cec613496d3a5e1b0891c9ff83e6/Lib/http/client.py#L1272) in the source code ([forcing latin1 encoding](https://github.com/python/cpython/blob/b00854caa080cec613496d3a5e1b0891c9ff83e6/Lib/http/client.py#L151)). Tried the following:

```python
In [1]: import http.client
In [2]: conn = http.client.HTTPSConnection("www.python.org")
In [3]: conn.request("GET", "/", headers={'X-utf-header': 'żółty'})

/usr/local/Cellar/python/3.6.4_4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/http/client.py in putheader(self, header, *values)
   1210         for i, one_value in enumerate(values):
   1211             if hasattr(one_value, 'encode'):
-> 1212                 values[i] = one_value.encode('latin-1')
   1213             elif isinstance(one_value, int):
   1214                 values[i] = str(one_value).encode('ascii')

UnicodeEncodeError: 'latin-1' codec can't encode character '\u017c' in position 0: ordinal not in range(256)
```

`curl` does support sending non-latin1 headers, but the behavior of servers is erratic and not consistent. I made a few tests. I tried both with a simple Flask app and a Django app; using their respective development servers and gunicorn. Both development servers drop completely the headers that can't be parsed. Gunicorn makes it available, but they're completely scrambled and are pretty much useless. These are the results of my experiments (using gunicorn; dev servers don't even make them available):

```bash
$ curl --header "X-header-ö21: jüźń" http://localhost:8000
# X-Special-Header: jÃ¼ÅºÅ
$ curl --header "X-ö21: jüźń" http://localhost:8000
# X-Header-Ã¶21: jÃ¼ÅºÅ
```

**Conclusion: ** I think this is way out of our scope. I haven't tried mod_wsgi or other variants, but I also expect some erratic behavior.

#### E2E tests

As @spookylukey mentioned, I'd love to do some E2E tests. When I implemented this solution I did write a few simple tests using [django-webtest](https://github.com/django-webtest/django-webtest) and a few others using `curl` and everything worked fine. Are there any "integration" tests in the test suite?

carltongibson commented:
@santiagobasulto Good follow-up! 

Lowercase Headers: hold-off for now to give people a chance to comment. (It's probably an _either way_ for most people.)

> Header Encoding... I think this is way out of our scope

I'm inclined to agree — **at least for this PR**. Ultimately all we have here is a filtered wrapper around the existing `META`, which we get given by WSGI. 

If we add nothing else we still have the API improvement. (Our position is certainly no worse.) _Maybe_ we can add other things later but I can't see that we should hold this up on that.

santiagobasulto commented:
Sorry for the spam, I was wondering, we could also rely on the [`SERVER_PROTOCOL`](http://wsgi.readthedocs.io/en/latest/definitions.html#envvar-SERVER_PROTOCOL) environ variable to check if it's `HTTP/2.0` and in that case make it all `lower()`.

collinanderson commented:
I think the most important thing is to be consistent, regardless of the underlying protocol. It doesn't actually go against the http2 spec if we were to use Title-Cased headers here, because we're not implementing http2.

santiagobasulto commented:
👋 hello folks. I've addressed all the comments made by @carltongibson. I made a small change to the `EnvironHeader` to make the "styling" of headers easily overridable. For now, they're still title-cased, but extending the class and changing it is extremely simple now (paving the way for future generations right? 😉).

@carltongibson I added the `versionadded` annotation as you suggested with version 2.1. Checking the [roadmap of 2.1](https://code.djangoproject.com/wiki/Version2.1Roadmap) I see there are still 3 days until it's frozen, so this one hopefully gets merged before that.

Thanks for your help and support! Have a great weekend y'all!

bluetech commented on docs/ref/request-response.txt:
```diff
@@ -159,14 +161,40 @@ All attributes should be considered read-only, unless stated otherwise.
+
+    .. versionadded:: 2.1
+
+    A case-insensitive dictionary containing the headers headers passed
```
"headers headers" should be "headers"

bluetech commented on docs/ref/request-response.txt:
```diff
@@ -159,14 +161,40 @@ All attributes should be considered read-only, unless stated otherwise.
+            'user-agent' in request.headers  # True
+
+            # getitem:
+            request.headers['User-Agent']  # Returns the value of Accept
```
Comments here and below should say "Returns the value of User-Agent".

bluetech commented on docs/ref/request-response.txt:
```diff
@@ -135,8 +135,10 @@ All attributes should be considered read-only, unless stated otherwise.
+    A dictionary containing the `WSGI environ \
+    <https://www.python.org/dev/peps/pep-0333/#environ-variables>`_.
+    Available request headers are prefixed with the string ``HTTP_`` and
+    depend on the client and server, but are some examples:
```
"but are" should say "but here are".

bluetech commented on django/utils/datastructures.py:
```diff
@@ -280,3 +282,94 @@ def __getitem__(self, key):
+        yield tuple(elem)
+
+
+def lowercased_key(method):
```
I don't think this decorator is particularly needed. I think the code would be clearer if you just do the `key.lower()` inside `__getitem__`.

bluetech commented on django/utils/datastructures.py:
```diff
@@ -280,3 +282,94 @@ def __getitem__(self, key):
+    def __repr__(self):
+        return repr({key: value for key, value in self._store.values()})
+
+    def copy(self):
```
If the type is indeed immutable, then `copy` can just be `return self`.

santiagobasulto replied:
I'd prefer a true copy of the object. With this method I make sure that `id(obj) != id(obj.copy())`. They're two different objects altogether.

bluetech replied:
Why do you need the IDs to be different? For example, consider an immutable collection from Python itself, `frozenset`. The following always prints True:

```
x = frozenset({1,2,3})
print(id(x) == id(x.copy()))
```

bluetech commented on django/http/request.py:
```diff
@@ -66,6 +68,12 @@ def __repr__(self):
             return '<%s>' % self.__class__.__name__
         return '<%s: %s %r>' % (self.__class__.__name__, self.method, self.get_full_path())
 
+    @property
```
You can use `django.utils.functional.cached_property` here and remove the custom caching code.

santiagobasulto replied:
Nice! I missed that one. I remember seeing it but then I completely forgot.

santiagobasulto commented:
Thanks @bluetech! I addressed all the comments. I am copying the dict, but I don't think it's a big issue compared to the simplicity and safety it adds to the implementation.

bluetech commented on django/utils/datastructures.py:
```diff
@@ -280,3 +281,86 @@ def __getitem__(self, key):
+        if not isinstance(other, Mapping):
+            return NotImplemented
+        return {
+            k.lower(): v for k, v in self.items()
```
The `lower()` call here can be removed, since you already know keys in `self.items()` are in lower case. In fact, you can just compare `self._store` instead.

santiagobasulto commented:
@carltongibson 👍 on it. I was away for a few days. I'll fix the comments and push. I should use release 2.2, right?

Stranger6667 commented on django/utils/datastructures.py:
```diff
@@ -280,3 +281,86 @@ def __getitem__(self, key):
+
+def _destruct_iterable_mapping_values(data):
+    for i, elem in enumerate(data):
+        if len(elem) != 2:
```
I believe, it can be avoided by explicit unpacking in `for`:
```
for i, (key, value) in enumerate(data):
    if not isinstance(key, str):
        raise ValueError('Element key invalid, only strings are allowed')
    yield (key, value)
```

The error message and behavior is the same.

timgraham commented:
Yes, target Django 2.2.

carltongibson commented on docs/releases/2.1.txt:
```diff
@@ -218,6 +218,10 @@ Signals
+* The new ``as_attachment`` argument for :class:`~django.http.FileResponse`
+  sets the ``Content-Disposition`` header to make the browser ask if the user
+  wants to download the file. ``FileResponse`` also tries to set the
+  ``Content-Type`` and ``Content-Length`` headers where appropriate.
```
It looks like you have an error rebasing here. (This isn't part of this PR.)

carltongibson commented on docs/ref/request-response.txt:
```diff
@@ -161,8 +162,9 @@ All attributes should be considered read-only, unless stated otherwise.
-    case-insensitive, hyphened headers (``request.headers['User-Agent']``)
+    ``HTTP_X_BENDER``.
+    See :attr:`~HttpRequest.headers` for more convenient header access
+    (``request.headers['User-Agent']``).
```
Re-wrap these lines please.

carltongibson commented on docs/ref/request-response.txt:
```diff
@@ -136,7 +136,8 @@ All attributes should be considered read-only, unless stated otherwise.
     A dictionary containing the `WSGI environ \
-    <https://www.python.org/dev/peps/pep-0333/#environ-variables>`_.
+    <https://www.python.org/dev/peps/pep-0333/#environ-variables>`_,
+    including all available HTTP headers.
```
Re-wrap this line with the one below.

carltongibson commented on docs/releases/2.1.txt:
```diff
@@ -207,6 +207,17 @@ Requests and Responses
+Signals
+~~~~~~~
+
+* ...
```
Remove this/rebase/put it 2.2.txt

santiagobasulto commented:
@carltongibson, thanks SO much for all your help. this community is amazing. I'll close the PR for a second so I can work on this without notifying everybody about changes and I'll reopen it after it's done.
Thanks again!
